src/dc_mimefactory.rs

```diff
@@ -83,6 +85,25 @@ impl<'a> MimeFactory<'a> {
             context,
         }
     }
+
+    pub fn finalize_mime_message(
+        &mut self,
+        message: *mut mailmime,
+        encrypted: bool,
+        gossiped: bool,
+    ) -> Result<(), Error> {
+        unsafe {
+            assert!(self.out.is_null()); // guard against double-calls
+            self.out = mmap_string_new(b"\x00" as *const u8 as *const libc::c_char);
+            let mut col: libc::c_int = 0;
+            if 0 != mailmime_write_mem(self.out, &mut col, message) {
+                bail!("could not write memory-mime-message")
+            }
+        }
+        self.out_encrypted = encrypted;
+        self.out_gossiped = encrypted && gossiped;
+        Ok(())
+    }
```

Comment:
this is the only function that ever serializes out a mime message, and it takes care to only be called once.

Reply:
I think you could make the `only call once` check even better by consuming `self` instead of using a reference. Not sure if this works out fully.

Reply:
left for further refactorings ;)